src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+/*! Standard Portable Intermediate Representation (SPIR-V) backend !*/
+use super::layout::{Instruction, Module};
+use crate::FastHashMap;
+use spirv::*;
+
+trait LookupHelper<T> {
+    type Target;
+    fn lookup_id(&self, handle: &crate::Handle<T>) -> Option<Word>;
+    fn lookup_handle(&self, word: &Word) -> Option<crate::Handle<T>>;
+}
+
+impl<T> LookupHelper<T> for FastHashMap<Word, crate::Handle<T>> {
+    type Target = T;
+
+    fn lookup_id(&self, handle: &crate::Handle<T>) -> Option<Word> {
+        let mut word = None;
+        for (k, v) in self.iter() {
+            if v.eq(handle) {
+                word = Some(*k);
+                break;
+            }
+        }
+        word
+    }
+
+    fn lookup_handle(&self, word: &u32) -> Option<crate::Handle<T>> {
+        let mut handle = None;
+        for (k, v) in self.iter() {
+            if k.eq(word) {
+                handle = Some(*v);
+                break;
+            }
+        }
+        handle
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct LookupFunctionType {
+    parameter_type_ids: Vec<Word>,
+    return_type_id: Word,
+}
+
+pub struct Parser {
+    module: Module,
+    id_count: u32,
+    capabilities: Vec<Capability>,
+    debugs: Vec<Instruction>,
+    annotations: Vec<Instruction>,
+    debug_enabled: bool,
+    void_type: Option<u32>,
+    lookup_type: FastHashMap<Word, crate::Handle<crate::Type>>,
+    lookup_function: FastHashMap<Word, crate::Handle<crate::Function>>,
+    lookup_function_type: FastHashMap<Word, LookupFunctionType>,
+    lookup_constant: FastHashMap<Word, crate::Handle<crate::Constant>>,
+    lookup_global_variable: FastHashMap<Word, crate::Handle<crate::GlobalVariable>>,
+    lookup_import: FastHashMap<Word, String>,
+    lookup_label: Vec<Word>,
+    current_label: Word,
+}
+
+impl Parser {
+    pub fn new(module: &crate::Module, debug_enabled: bool) -> Self {
```

Comment:
Boolean arguments are an anti-pattern. Perhaps, it could be just a public field?

Reply:
Could you elaborate why it is an anti-pattern and how you would like to see it as public field?

Reply:
Simply because when you see a function call `foo(something, true)`, it doesn't give you any idea on what that `true` means. You could say the same thing about integers, but the problem with `bool` is that it's notoriously general and used often.

The idiomatic way to do it would be either having
```rust
enum Debug {
  On,
  Off,
}
```
so that the call spot always sees the semantics.
Or having it a bitflag: `ParserFlags::DEBUG`
Or having it the builder pattern (or an init struct): `Parser::new(module).with_debug()`.

Basically, any API where the meaning of this is obvious at the call site :)